### PR TITLE
executor: fix bug of point get when meet null values

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

For point get queries like `select b, b from t where a = 1`, if `b` is null, tidb will panic now.

### What is changed and how it works?

The reason is that `len(decodedVals[decodedPos])` will be 0 for null values, then it will not make reference for the second columns in the chunk, so it will panic when tidb tries to produce the result. This PR makes reference in the chunk for it.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: executor/point_get.go [hunk 1]
```diff
@@ -221,19 +221,13 @@ func (e *PointGetExecutor) decodeRowValToChunk(rowVal []byte, chk *chunk.Chunk)
...
-			}
 			continue
 		}
-		// ExtraHandleID is added when building plan, we can make sure that there's only one column's ID is this.
 		if id == model.ExtraHandleID {
 			chk.AppendInt64(firstPos, e.handle)
 			continue
 		}
 		if len(decodedVals[decodedPos]) == 0 {
-			// This branch only entered for updating and deleting. It won't have one column in multiple positions.
```
Comment (winoros) [inline executor/point_get.go]: This assumption is broken?

Comment (lamxTyler) [reply executor/point_get.go]: Yes, it could be null.

**Comment by coocood** [comment]
This function looks complex, can we remove `decodedPos2SchemaPos` and range over `e.schema.Columns` to decode to chunk?

**Comment by winoros** [comment]
@coocood If we remove it, we still need to implement the same logic like this map since the column value can only be decoded once but it can be referenced multiple times?

**Comment by coocood** [comment]
@winoros 
We can use `col.ID` to find the position in `colID2DecodedPos`, then get the value from the decoded row.

**Comment by lamxTyler** [comment]
@coocood I think the main point of @winoros is if we need to decode one column several times, if not, then the code will not be cleaner. By the way, even if we can decode it several times, it does not make it cleaner because most of the codes are dealing with the different situation of columns.

**Comment by coocood** [comment]
@lamxTyler 
Just decode as needed is straight forward, we don't need to optimize this case, it's a very rare case.

**Comment by coocood** [comment]
I think this is much simpler
```go
func (e *PointGetExecutor) decodeRowValToChunk(rowVal []byte, chk *chunk.Chunk) error {
	//  One column could be filled for multi-times in the schema. e.g. select b, b, c, c from t where a = 1.
	// We need to set the positions in the schema for the same column.
	colID2DecodedPos := make(map[int64]int, e.schema.Len())
	for _, col := range e.schema.Columns {
		if _, ok := colID2DecodedPos[col.ID]; !ok {
			colID2DecodedPos[col.ID] = len(colID2DecodedPos)
		}
	}
	cutVals, err := tablecodec.CutRowNew(rowVal, colID2DecodedPos)
	if err != nil {
		return err
	}
	if cutVals == nil {
		cutVals = make([][]byte, len(colID2DecodedPos))
	}
	decoder := codec.NewDecoder(chk, e.ctx.GetSessionVars().Location())
	for colIdx, col := range e.schema.Columns {
		if e.tblInfo.PKIsHandle && mysql.HasPriKeyFlag(col.RetType.Flag) {
			chk.AppendInt64(colIdx, e.handle)
			continue
		}
		// ExtraHandleID is added when building plan, we can make sure that there's only one column's ID is this.
		if col.ID == model.ExtraHandleID {
			chk.AppendInt64(colIdx, e.handle)
			continue
		}
		cutVal := cutVals[colID2DecodedPos[col.ID]]
		if cutVal == nil {
			colInfo := getColInfoByID(e.tblInfo, col.ID)
			d, err1 := table.GetColOriginDefaultValue(e.ctx, colInfo)
			if err1 != nil {
				return err1
			}
			chk.AppendDatum(colIdx, &d)
			continue
		}
		_, err = decoder.DecodeOne(cutVal, colIdx, col.RetType)
		if err != nil {
			return err
		}
	}
	return nil
}
```

File: executor/point_get.go [hunk 2]
```diff
@@ -196,16 +196,10 @@ func (e *PointGetExecutor) get(key kv.Key) (val []byte, err error) {
 }
 
 func (e *PointGetExecutor) decodeRowValToChunk(rowVal []byte, chk *chunk.Chunk) error {
-	//  One column could be filled for multi-times in the schema. e.g. select b, b, c, c from t where a = 1.
-	// We need to set the positions in the schema for the same column.
 	colID2DecodedPos := make(map[int64]int, e.schema.Len())
```
Comment (coocood) [inline executor/point_get.go]: The value is only cut, not decoded.
`colID2CutPos` is better.

File: executor/point_get.go [hunk 3]
```diff
@@ -196,16 +196,10 @@ func (e *PointGetExecutor) get(key kv.Key) (val []byte, err error) {
...
-		if decodedPos, ok := colID2DecodedPos[col.ID]; !ok {
+	for _, col := range e.schema.Columns {
+		if _, ok := colID2DecodedPos[col.ID]; !ok {
 			colID2DecodedPos[col.ID] = len(colID2DecodedPos)
-			decodedPos2SchemaPos = append(decodedPos2SchemaPos, []int{schemaPos})
-		} else {
-			decodedPos2SchemaPos[decodedPos] = append(decodedPos2SchemaPos[decodedPos], schemaPos)
 		}
 	}
 	decodedVals, err := tablecodec.CutRowNew(rowVal, colID2DecodedPos)
```
Comment (coocood) [inline executor/point_get.go]: cutVals

**Comment by coocood** [comment]
LGTM

**Pull request merged**